### DISCUSSION TITLE
Use of `npx`

### DISCUSSION THREAD

**Discussion by ghost** [xpack/vscode-xpack-extension-ts#46] [Ideas]
Is it a good idea to integrate the possibility to use `npx` if `xpm` is installed as project's Node module (`<project_root>/node_modules/`)?

> ℹ️ For me, it's a solution to execute `xpm` within a Windows with company policies restrictions.

`xpm` is a NodeJS dependency of the project defined within the _package.json_ file:
```json
  "dependencies": {
    "xpm": "^0.16.1"
  },
```

Its installation is done with the following command executed from project root: `npm install`

To use it, we should be call `xpm` through `npx`. For isntance: `npx xpm install`

This ask to offer the possibility to overload the VS Code xPack actions.

**Comment by ilg-ul** [comment]
> it's a solution to execute xpm within a Windows with company policies restrictions.

I'm not sure I understand your use case. What are those restrictions? 

> ... the possibility to overload the VS Code xPack actions.

I definitely don't understand this. Please elaborate.

**Comment by ghost** [comment]
First, sorry for my english... and thank you a lot for your responsiveness.

Unfortunately, my company forbids to execute a program/script from `C:\Users\<user_name>\AppData\Roaming` without admin privileges...

But my company authorizes to execute a program/script from a project path, therefore I installed `xpm` as a project's Node module `(<project_root>/node_modules/xpm)`
`xpm` is a NodeJS dependency of the project defined within the _package.json_ file:
```json
  "dependencies": {
    "xpm": "^0.16.1"
  },
```

I install `xpm` with `npm install`. And to avoid update my Windows `path` environment variable to define the path of project's `xpm` I use [NPX](https://stackoverflow.com/questions/9679932/how-to-use-executables-from-a-package-installed-locally-in-node-modules).

Therefore I created VS Code tasks to overload the VS Code xPacks actions to add `npx` in front of the VS Code xPacks action commands.

For instance:
```json
    {
      "label": "Clean (copy of xpack action)",
      "type": "shell",
      "command": "npx xpm run clean --config Debug -C '${workspaceRoot}'",
      "problemMatcher": []
    },
```

My suggestion is to add this `npx`to the VS Code xPacks actions if _package.json_ file contains `xpm` as a dependency.

I hope that is more understandable. 🤞

**Comment by ilg-ul** [comment]
> my company forbids to execute a program/script from C:\Users\<user_name>\AppData\Roaming without admin privileges...

I see... Not good... :-(

Can you define some environment variables? You can choose a different location for the xPacks:

- https://xpack.github.io/xpm/folders/

**Comment by ghost** [comment]
> Can you define some environment variables? You can choose a different location for the xPacks:

Yes, I can create environment variables but the installation and using of xPacks through the symbolic links of _xpacks_ folder within the project work fine.
The ban is only on the execution of xpm from from `C:\Users\<user_name>\AppData\Roaming\npm`...

```pwsh
❯ where.exe xpm
C:\Users\<user_name>\AppData\Roaming\npm\xpm
C:\Users\<user_name>\AppData\Roaming\npm\xpm.cmd
```

**Reply by ilg-ul** [reply]
> The ban is only on the execution of xpm from from C:\Users\<user_name>\AppData\Roaming\npm...

Ah, so the restriction is on all npm global packages. I'm afraid there is not much to do in this case. 

However, for Windows, the nodeJs is also distributed as a .zip archive, which I expect to use a custom location. Did you try it?

Using npx may solve some specific cases, but it is not a general solution, many xPack actions cascade other actions, and so on.

**Reply by ghost** [reply]
> However, for Windows, the nodeJs is also distributed as a .zip archive, which I expect to use a custom location. Did you try it?

NodeJS with its NPM? No, I never tried to install it with its .zip archive.

> [...] many xPack actions cascade other actions, and so on.

A xPack action starts always by `xpm run <action>`, right?

> Using npx may solve some specific cases, but it is not a general solution

I understand your opinion.

Again, thank you for your great job my friend of Mera Peak 😄 (I'm reached the summit also in 2019)
